File: artifacts/heman/decompose/src/import.c
heman_image *heman_import_u8(int width, int height, int nbands, const heman_byte *source, float minval, float maxval)
{
  heman_image *result = heman_image_create(width, height, nbands);
  const heman_byte *inp = source;
  unsigned int inp_idx = 0;
  float *outp = result->data;
  float scale = (maxval - minval) / 255.0f;
  int size = (height * width) * nbands;
  for (int i = 0; i < size; i += 1)
  {
    float v = (inp[inp_idx] * scale) + minval;
    inp_idx += 1;
    *(outp++) = (minval > ((maxval > v) ? (v) : (maxval))) ? (minval) : ((maxval > v) ? (v) : (maxval));
  }

  return result;
}


/*** DEPENDENCIES:
heman_image *heman_image_create(int width, int height, int nbands)
{
  heman_image *img = malloc(sizeof(heman_image));
  img->width = width;
  img->height = height;
  img->nbands = nbands;
  img->data = malloc((((sizeof(float)) * width) * height) * nbands);
  return img;
}


----------------------------
typedef unsigned char heman_byte
----------------------------
typedef struct heman_image_s heman_image
----------------------------
***/

File: artifacts/heman/translate_without_break_down/src/import.rs
use crate::*;
use rand::Rng;


pub fn heman_import_u8(
    width: i32,
    height: i32,
    nbands: i32,
    source: Option<&[heman_byte]>,
    minval: f32,
    maxval: f32,
) -> HemanImage {
    // Check for NULL equivalent (None in Rust)
    let source = match source {
        Some(s) => s,
        None => return None,
    };

    // Create the result image
    let mut result = match heman_image_create(width, height, nbands) {
        Some(r) => r,
        None => return None,
    };

    // Calculate scale factor and total size
    let scale = (maxval - minval) / 255.0f32;
    let size = (height * width * nbands) as usize;

    // Ensure we have enough data in source
    if source.len() < size {
        return None;
    }

    // Process the data
    if let Some(ref mut data) = result.data {
        for i in 0..size {
            let v = (f32::from(source[i]) * scale) + minval;
            // Clamp the value between minval and maxval
            data[i] = if v < minval {
                minval
            } else if v > maxval {
                maxval
            } else {
                v
            };
        }
    }

    Some(result)
}
